### PR TITLE
FORGE-1525: jpa-new-entity: Allow @EmbeddedId

### PR DESCRIPTION
**Pull request by criztovyl**
This PR is not ready yet, but I submitted it early to get some feedback. :)
Corresponding JBoss Issue is [FORGE-1525](https://issues.jboss.org/browse/FORGE-1525).

My thoughts are adding `--id-type` to `jpa-new-entity` to allow using other ID types than the default `@Id Long id;`, e.g. `--id-type EMBEDDED_ID`:
```java
@EmbeddedId
MyEmbeddedId embeddedId;
```
or `--id-type ID_CLASS`
```java
@IdClass(MyId.class) class MyEntity { }
```

The Id's class for the embedded id currently is set via `--id-class`, but for `@IdClass` with multiple properties needs to be handled differently if you also want to set the property names.

I've already implemented `@EmbeddedId` roughly, and I want to continue with `@IdClass`.

Code is not perfect yet, I'm currently warming up with Forge.

### CODE REVIEW

File: javaee/impl/src/main/java/org/jboss/forge/addon/javaee/jpa/ui/JPANewEntityCommandImpl.java [hunk 1]
```diff
@@ -86,13 +108,31 @@ public JavaClassSource decorateSource(UIExecutionContext context, Project projec
...
+
+      switch(idTypeChosen){
+          case LONG_PROPERTY:
+              return persistenceOperations.newEntity(source, idStrategyChosen, tableName.getValue(), idPropertyName,
+                      versionPropertyName);
+          case EMBEDDED_ID:
+              // XXX what if idClass value is empty?
+              return persistenceOperations.newEntityEmbeddedId(source, tableName.getValue(), idPropertyName,
+                      idClass.getValue(), versionPropertyName);
+          case ID_CLASS:
```
Comment (gastaldi) [inline javaee/impl/src/main/java/org/jboss/forge/addon/javaee/jpa/ui/JPANewEntityCommandImpl.java]: In this case perhaps we should remove from the enum?

File: javaee/impl/src/main/java/org/jboss/forge/addon/javaee/jpa/ui/JPANewEntityCommandImpl.java [hunk 2]
```diff
@@ -74,7 +86,17 @@ public void initializeUI(UIBuilder builder) throws Exception
    {
       super.initializeUI(builder);
       idStrategy.setDefaultValue(GenerationType.AUTO);
-      builder.add(idStrategy).add(tableName);
+      idType.setDefaultValue(EntityIdType.LONG_PROPERTY);
+      idClass.setRequired(new Callable<Boolean>(){
```
Comment (gastaldi) [inline javaee/impl/src/main/java/org/jboss/forge/addon/javaee/jpa/ui/JPANewEntityCommandImpl.java]: You can use a lambda here:`idClass.setRequired(() -> idType.getValue() == EntityIdType.EMBEDDED_ID);`

**Comment by gastaldi** [comment]
Don't forget to add tests too ;)

**Comment by gastaldi** [comment]
Retest this please

**Comment by gastaldi** [comment]
Ok to test

File: javaee/impl/src/main/java/org/jboss/forge/addon/javaee/jpa/ui/JPANewEntityCommandImpl.java (hunk 1 above)
Comment (criztovyl) [reply javaee/impl/src/main/java/org/jboss/forge/addon/javaee/jpa/ui/JPANewEntityCommandImpl.java]: `idClass` and `ID_CLASS` are unrelated.
`idClass` is the Class/Type to use for the `@EmbeddedId`, `ID_CLASS` means to use the `@IdClass`-Annotation (at the Entity), in contrast to `EMBEDDED_ID` which means `@EmbeddedId` (at the property).

Or Do you mean the `UnsupportedOperationException`?
That one is only temporary until I worked out how I want to do `@IdClass`.

File: javaee/impl/src/main/java/org/jboss/forge/addon/javaee/jpa/ui/JPANewEntityCommandImpl.java (hunk 2 above)
Comment (criztovyl) [reply javaee/impl/src/main/java/org/jboss/forge/addon/javaee/jpa/ui/JPANewEntityCommandImpl.java]: :+1:

File: javaee/impl/src/main/java/org/jboss/forge/addon/javaee/jpa/ui/JPANewEntityCommandImpl.java (hunk 1 above)
Comment (gastaldi) [reply javaee/impl/src/main/java/org/jboss/forge/addon/javaee/jpa/ui/JPANewEntityCommandImpl.java]: Ah right, I thought that wasn't temporary, that's fine :+1:

**Comment by gastaldi** [comment]
add to whitelist

**Comment by criztovyl** [comment]
Is the `@EmbeddedId` test is okay that way?

I've added the `@IdClass` strategy now, too.
It works by locating all fields on the class given via `--id-class` and adding them and the `@IdClass ` annotation to the new entity class.
But I don't know yet how to get hands onto an `JavaClassSource` or similar.

There also is a test, but it's not tested yet, so won't push it for now.

**Comment by gastaldi** [comment]
A `JavaClassSource` is part of the Roaster library. See how to use it here: https://github.com/forge/roaster/blob/master/README.asciidoc#java-parser-api

You can use the JavaSourceFacet to find a specific class in your project. See http://forge.jboss.org/addon/org.jboss.forge.addon:parser-java and https://github.com/forge/core/blob/master/parser-java/api/src/main/java/org/jboss/forge/addon/parser/java/facets/JavaSourceFacet.java for more info.

File: javaee/impl/src/main/java/org/jboss/forge/addon/javaee/jpa/PersistenceOperationsImpl.java [hunk 3]
```diff
@@ -181,6 +186,40 @@ public JavaClassSource newEntity(JavaClassSource javaClass, GenerationType idStr
...
+               .setLiteralValue("nullable", "false");
+
+       return newEntity(javaClass, tableName, id, versionPropertyName);
+   }
+
+   @Override
+   public JavaClassSource newEntityEmbeddedId(JavaClassSource javaClass, String tableName, String idPropertyName,
+           String idPropertyType, String versionPropertyName)
+   {
+       // XXX what if the given class ist not @Embeddable?
```
Comment (gastaldi) [inline javaee/impl/src/main/java/org/jboss/forge/addon/javaee/jpa/PersistenceOperationsImpl.java]: Add the @Embdeddable annotation or throw an Exception?

Comment (criztovyl) [reply javaee/impl/src/main/java/org/jboss/forge/addon/javaee/jpa/PersistenceOperationsImpl.java]: `XXX` as in `TODO` here, maybe I messed sth up. ^^
I wIll thrown an exception, modifying existing classes seems wrong here.

**Comment by criztovyl** [comment]
It seems like I got it working now. \o/

File: javaee/impl/src/main/java/org/jboss/forge/addon/javaee/jpa/ui/JPANewEntityCommandImpl.java [hunk 4]
```diff
@@ -86,13 +108,32 @@ public JavaClassSource decorateSource(UIExecutionContext context, Project projec
...
+                      versionPropertyName);
+          case EMBEDDED_ID:
+              // XXX what if idClass value is empty?
+              return persistenceOperations.newEntityEmbeddedId(source, tableName.getValue(), idPropertyName,
+                      idClass.getValue(), versionPropertyName);
+          case ID_CLASS:
+              return persistenceOperations.newEntityIdClass(source, tableName.getValue(), idClass.getValue(),
+                      versionPropertyName);
+          default:
+              throw new IllegalArgumentException("Unknow Enum value " + idTypeChosen + "!");
```
Comment (gastaldi) [inline javaee/impl/src/main/java/org/jboss/forge/addon/javaee/jpa/ui/JPANewEntityCommandImpl.java]: Typo in `Unknown`

File: javaee/impl/src/main/java/org/jboss/forge/addon/javaee/jpa/ui/JPANewEntityCommandImpl.java [hunk 5]
```diff
@@ -86,13 +105,43 @@ public JavaClassSource decorateSource(UIExecutionContext context, Project projec
...
-               versionPropertyName);
+
+      JavaResource jr = project.getFacet(JavaSourceFacet.class).getJavaResource(idClass.getValue());
+
+      if(!(jr.getJavaType() instanceof JavaClassSource)){
+          throw new IllegalArgumentException("The found JavaResource for the ID class is not a JavaClassResource, " +
+                  "which is required. Found resource: " + jr.getClass().getCanonicalName());
+      }
+
+      JavaClassSource jcs = (JavaClassSource) jr.getJavaType();
```
Comment (criztovyl) [inline javaee/impl/src/main/java/org/jboss/forge/addon/javaee/jpa/ui/JPANewEntityCommandImpl.java]: Uhm, this breaks if `LONG_PROPERTY` should be used... :D

Comment (criztovyl) [reply javaee/impl/src/main/java/org/jboss/forge/addon/javaee/jpa/ui/JPANewEntityCommandImpl.java]: I'll put it into an own function that is called in the `ID_CLASS` and `EMBEDDED_ID` case below.

File: javaee/tests/src/test/java/org/jboss/forge/addon/javaee/jpa/ui/JPANewEntityCommandTest.java [hunk 6]
```diff
@@ -128,4 +132,32 @@ public void checkCommandShell() throws Exception
...
    }
+
+   @SuppressWarnings("unchecked")
+   @Test
+   public void checkCommandShellEmbeddedId() throws Exception
+   {
+      shellTest.getShell().setCurrentResource(project.getRoot());
+
+      Result result = shellTest
+               .execute("jpa-new-entity --named Customer --target-package space.criztovyl --id-type EMBEDDED_ID --id-class space.criztovyl.CustomerId",
```
Comment (criztovyl) [inline javaee/tests/src/test/java/org/jboss/forge/addon/javaee/jpa/ui/JPANewEntityCommandTest.java]: As I now check for `@Embedded`, I will need to create the Id class beforehand.

**Comment by criztovyl** [comment]
So far, so good.

There still is an uncaught exception when the selected Id-Class does not exist, I will try to catch it.

After the above fix, the PR would be nearly ready from my side.
With your approval I would complete it by rebasing my branch to merge the CI-fix commits into one while keeping the overall structure.
I would prefer not to squash all my commits - I like to keep them somewhat atomic - if you could merge my branch using a merge-commit, that would make me really happy. :)

Thanks.

**Comment by gastaldi** [comment]
Retest this please

**Comment by gastaldi** [comment]
Hey @criztovyl! The changes look really great, good job! I have two remarks: 

1) Some classes are not using the [Forge Code Formatter](https://github.com/forge/core/blob/master/eclipse-code-formatter-profile.xml). I think that's probably my fault for not making that information explicit in the [Contribute Code](http://forge.jboss.org/document/contribute-code) page;

2) Unfortunately we never do merge commits in our core repository. It´s a guideline we follow for some years, and we always squash commits in PRs to a single commit (that is easier for us to detect changes and revert once something goes wrong. Hope that doesn't stop you from contributing, we really appreciate your work :) ) 

Best Regards,

George Gastaldi

File: javaee/api/src/main/java/org/jboss/forge/addon/javaee/jpa/EntityIdType.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
+/**
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates.
```
Comment (gastaldi) [inline javaee/api/src/main/java/org/jboss/forge/addon/javaee/jpa/EntityIdType.java]: This can be 2018 ;)

**Comment by criztovyl** [comment]
1. I've read about the formatting in the contribution guide, but forgot to follow it, so that's rather my fault. :)
(I'm not using Eclipse nor IDEA, but I'll find a way to do it with the given settings somehow non-manually. ^^)

2. I understand that, it's okay for me, but I'll be "sad" now :P

I have some things to do this week, I'll come back here on friday or on the weekend, just fyi.

**Comment by gastaldi** [comment]
Sure, no rush. Again, thank you very much for your contribution, we really appreciated that!

**Comment by criztovyl** [comment]
I, in return, appreciate your appreciation, thanks. :)

**Comment by criztovyl** [comment]
I did not take any time for this PR this weekend, I'll continue later.
Maybe I'll find time earlier, but I think it will take until end of the month.

btw, on my journey to my merge commit, I found out that reverting a non-ff merge-commit is nearly as easy as reverting a squashed merge-commit: [Undoing merges](https://git-scm.com/blog/2010/03/02/undoing-merges.html) (the second part).
That way it would even be very easy to revert only parts of my contribution, e.g. if I, hypothetically, would have messed up the `@IdClass` case or something like that.
(I really, really like merge commits, sry to bother you with that ^^)

**Comment by gastaldi** [comment]
That's fine. Just remove the [WIP] in this PR when you feel it should be merged

**Comment by criztovyl** [comment]
<sup>*Not a the month as originally said, but nobody knows that.* :zap:</sup>

**Comment by gastaldi** [comment]
Retest this please

**Comment by gastaldi** [comment]
Retest this please

**Comment by gastaldi** [comment]
Thanks!

**Pull request merged**